### PR TITLE
[ACS] add AKS (managed Kubernetes clusters)

### LINKED ISSUE #1386: [Container Service]: Add Swagger for `2017-07-01`
👋 hey y'all

The [Azure Container Service recently announced support for the `2017-07-01` API](https://github.com/Azure/ACS/blob/master/announcements/2017-06-28-acs-uk-public-preview.md) - which solves a ton of issues we've raised previously, including:
 - #833 - Windows support
 - #1172 - Internal Clusters
 - https://github.com/terraform-providers/terraform-provider-azurerm/issues/79 - Tearing down Azure ACS resource does not tear down resource group
 - https://github.com/terraform-providers/terraform-provider-azurerm/issues/45 - Multiple Agent Pools

Would it be possible to add the Swagger file associated with this API version?

Thanks!

### PR DESCRIPTION
**Pull request by mboersma**
Closes #1386.

This replaces #1912 with an updated API for ACS/AKS which lives in its own specification directory and uses `x-ms-client-flatten` consistently.

ping @rite2nikhil @shrutir25 @rjtsdl

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [x] If applicable, the PR references the bug/issue that it fixes.
- [x] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [x] My spec meets the review criteria:
  - [x] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [x] The spec follows the guidelines described in the [Swagger checklist](../documentation/swagger-checklist.md) document.
  - [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

### CODE REVIEW

**Comment by mboersma** [comment]
Here is a zipped version of the azure-mgmt-containerservice 2.1.0 .whl if it helps anyone in testing these changes with the `az` CLI:
 [azure_mgmt_containerservice-2.1.0-py2.py3-none-any.whl.zip](https://github.com/Azure/azure-rest-api-specs/files/1434660/azure_mgmt_containerservice-2.1.0-py2.py3-none-any.whl.zip)

File: specification/containerservices/resource-manager/readme.md
Comment (lmazuel) [inline specification/containerservices/resource-manager/readme.md]: This tag "package-2017-01 and python" is not true anymore, if you could save me a PR and do it at the same time :)

**Comment by mboersma** [comment]
@dsgouda could we please get this reviewed today? I'm concerned that if this doesn't merge soon, we have no chance of making the necessary `az` CLI changes before the next ship deadline on November 8.

- `az aks` went live on Oct. 24 and has been popular beyond our wildest dreams, but its API definition is still not merged publicly, which hampers SDK and docs generation
- this should be a stable API going forward into GA
- its already had many eyes on it in private preview review and in #1912
- all feedback previously given has been addressed

**Comment by dsgouda** [comment]
Apologize for the delay, I missed the notification that I was assigned this. Will review it right away.

File: specification/containerservices/resource-manager/Microsoft.ContainerService/2017-07-01/containerService.json
Comment (dsgouda) [inline specification/containerservices/resource-manager/Microsoft.ContainerService/2017-07-01/containerService.json]: Making sure you understand the risks here, setting `modelAsString` to `false` will cause breaking changes in sdks whenever the enum is modified. Please acknowledge.

File: specification/containerservices/resource-manager/Microsoft.ContainerService/2017-07-01/containerService.json
Comment (dsgouda) [inline specification/containerservices/resource-manager/Microsoft.ContainerService/2017-07-01/containerService.json]: Making sure you understand the risks here, setting modelAsString to false will cause breaking changes in sdks whenever the enum is modified. Please acknowledge.

File: specification/containerservices/resource-manager/Microsoft.ContainerService/2017-07-01/containerService.json
Comment (dsgouda) [inline specification/containerservices/resource-manager/Microsoft.ContainerService/2017-07-01/containerService.json]: Please mark this as readonly

File: specification/containerservices/resource-manager/Microsoft.ContainerService/2017-08-31/managedClusters.json
Comment (dsgouda) [inline specification/containerservices/resource-manager/Microsoft.ContainerService/2017-08-31/managedClusters.json]: Same comment as above for `modelAsString`

File: specification/compute/resource-manager/readme.md
Comment (dsgouda) [inline specification/compute/resource-manager/readme.md]: Just want to know the intention here, are you sure you want to delete compute-specific tags even though most changes here are related to `ContainerServices`

File: specification/containerservices/resource-manager/readme.md
Comment (dsgouda) [inline specification/containerservices/resource-manager/readme.md]: Is the intention here to package the bits along with the compute ones.

File: specification/containerservices/resource-manager/Microsoft.ContainerService/2017-09-30/location.json
Comment (dsgouda) [inline specification/containerservices/resource-manager/Microsoft.ContainerService/2017-09-30/location.json]: Looks like a resource, must be modeled as a resource

**Review by dsgouda — CHANGES_REQUESTED** [review]
For the most part looks good.  A few comments and questions need to be addressed, errors flagged by the validator must also be addressed.

File: specification/containerservices/resource-manager/Microsoft.ContainerService/2017-08-31/managedClusters.json
Comment (dsgouda) [inline specification/containerservices/resource-manager/Microsoft.ContainerService/2017-08-31/managedClusters.json]: Mark this as readonly

File: specification/compute/resource-manager/readme.md
Comment (mboersma) [reply specification/compute/resource-manager/readme.md]: This particular tag is identical to the `package-2017-03` tag above it once `containerService.json` was moved, so I removed it.

File: specification/containerservices/resource-manager/readme.md
Comment (mboersma) [reply specification/containerservices/resource-manager/readme.md]: The containerServices API was previously part of the compute package, but this PR separates it. If someone needs to regenerate the Go SDK with `autorest --tag=package-2017-01` this should produce the same results as before, which seemed to me like the safe approach.

But if you'd rather have the paths changed to say `containerservice` or have these sections removed, let me know.

File: specification/containerservices/resource-manager/Microsoft.ContainerService/2017-07-01/containerService.json
Comment (mboersma) [reply specification/containerservices/resource-manager/Microsoft.ContainerService/2017-07-01/containerService.json]: Hadn't thought of that, but good point. I don't see why we can't use `"modelAsString": true` here and elsewhere, so I'll make that change.

File: specification/containerservices/resource-manager/Microsoft.ContainerService/2017-07-01/containerService.json
Comment (mboersma) [reply specification/containerservices/resource-manager/Microsoft.ContainerService/2017-07-01/containerService.json]: Acknowledged: this value is unlikely to change in the future, and we want to benefit from the [validation](https://github.com/Azure/azure-rest-api-specs/blob/current/documentation/creating-swagger.md#modelasstring) that happens with `"modelAsString": false`.

File: specification/containerservices/resource-manager/Microsoft.ContainerService/2017-08-31/managedClusters.json
Comment (mboersma) [reply specification/containerservices/resource-manager/Microsoft.ContainerService/2017-08-31/managedClusters.json]: Same response as above. 😄

File: specification/containerservices/resource-manager/Microsoft.ContainerService/2017-07-01/containerService.json
Comment (mboersma) [reply specification/containerservices/resource-manager/Microsoft.ContainerService/2017-07-01/containerService.json]: Done.

File: specification/containerservices/resource-manager/readme.md
Comment (dsgouda) [reply specification/containerservices/resource-manager/readme.md]: Ahh I see, so I am guessing new packages will be generated at a different path, that sounds okay

File: specification/containerservices/resource-manager/Microsoft.ContainerService/2017-07-01/containerService.json
Comment (dsgouda) [reply specification/containerservices/resource-manager/Microsoft.ContainerService/2017-07-01/containerService.json]: As due diligence we like to point it out in PRs, reduces future pains. 😄

**Comment by mboersma** [comment]
Travis CI seems to be failing this for reasons unrelated to this PR:
```
 1) Azure swagger schema validation:
       /home/travis/build/Azure/azure-rest-api-specs/specification/automation/resource-manager/Microsoft.Automation/2017-05-15-preview/softwareUpdateConfiguration.json should be a valid Swagger document.:
     Error: Schema validation failed: [ { code: 'ONE_OF_MISSING',
    params: [],
    message: 'Data does not match any schemas from \'oneOf\'',
    path: 
```

**Comment by dsgouda** [comment]
Yup, investigating. Also have you gone through the validation errors flagged here.

**Comment by dsgouda** [comment]
Please address the validation errors flagged [here](https://travis-ci.org/Azure/azure-rest-api-specs/jobs/298783525)

**Comment by mboersma** [comment]
Some of the validation errors are not going to be fixed soon, if at all--ACS/AKS doesn't have any PATCH endpoints, for example, and a previous reviewer advised us the operations endpoint error could be ignored. But we're trying to clarify what needs to be done here--I will update soon. 

Until we know how to proceed on some of those errors, we should consider this a work-in-progress. We need to get this in as soon as it's ready, but it's not going to merge today obviously, and the CLI refactoring related to this can wait a bit.

**Comment by dsgouda** [comment]
Noted. While I'm okay with holding off on other errors for now, you `must` expose an Operations API that return information related to the service.
@veronicagg just want to confirm if we can ignore the errors here.

**Comment by veronicagg** [comment]
OperationAPI violation is an RPC violation, so it's really a question for @ravbhatnagar :)

**Comment by dsgouda** [comment]
@ravbhatnagar there are no new service endpoints as such being introduced but this does add a bunch of errors in the reporting, do you think you can do a quick signoff here?

**Comment by ravbhatnagar** [comment]
@mboersma - operations API is required for the RP as per the RPC. Please find the details here. Your RP will have to expose this API. https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/proxy-api-reference.md#exposing-available-operations

**Comment by StrongMonkey** [comment]
when will this pr get in? We are waiting for the go-sdk library to integrate AKS into our own tool.

**Comment by matyix** [comment]
@StrongMonkey until this is merged I've built my own Go SDK for AKS (just cleaning and wrapping it up) - see [here](https://github.com/banzaicloud/azure-aks-client) but +1 for the PR merge ... @dsgouda @salameer @ravbhatnagar - can we help anyhow (with testing, etc) to get this merged in the near future?

**Comment by ravbhatnagar** [comment]
We can merge it. But please note this is the API that is used for RBAC, creating custom RBAC roles, activity logs logging, Having operations/actions exposed through the /operations API is a KPI that we track and it gets reported in the monthly executive review. Having this API on the service is extremely important. So we can still merge this PR, but please consider the above mentioned cases. @salameer  as FYI.

**Comment by dsgouda** [comment]
@StrongMonkey you need to resolve the merge conflicts. I will approve at my end and @salameer  can find someone to merge once the conflicts are resolved.
Also, please acknowledge the points @ravbhatnagar raised.

**Review by dsgouda — APPROVED** [review]
LGTM pending merge conflicts resolution

**Comment by StrongMonkey** [comment]
@dsgouda i am not the contributor of the original pr. Could @mboersma possibly solve the conflicts?

**Comment by lmazuel** [comment]
@devigned This was conflicts I introduced part of my work on Compute 2017-12-01. so I fixed them.

**Comment by hyonholee** [comment]
Thanks for separating Container service APIs from Compute.  Please also remove them from Compute in .NET client library and Powershell.

https://github.com/Azure/azure-sdk-for-net/tree/psSdkJson6/src/SDKs/Compute/Management.Compute/Generated

https://github.com/Azure/azure-powershell/tree/preview/src/ResourceManager/Compute/Commands.Compute/Generated/ContainerService

**Comment by devigned** [comment]
@lmazuel thank you!

@shahabhijeet can you coordinate the removal / separation from Compute via @hyonholee comment above.

**Comment by hyonholee** [comment]
Hi, please do not remove Container service Swagger specs from compute/resource-manager, until AKS client library is published to Nuget.  Currently, Azure Powershell Compute module contains container service cmdlets.  We need to keep these cmdlets until AKS team publishes its own Azure Powershell module.   Thus, Compute client library should include container service APIs, until a separate AKS client library is published.  So I suggest divide this PR to two separate stages.

Stage 1. add only files under containserservice/resource-manager folder, and let this PR merged.

After Stage 1, Publish AKS client library to nuget with 2017-01-31 version (The APIs of this client library should be identical as the container service APIs of the current Compute client library.  (say, this is AKS version 1.0)

Then, Publish AKS client library to nuget with the latest version (say, this is AKS version 2.0)

Stage 2. send another PR for removing container service specs from compute/resource-manager.

In this way, Azure Powershell Compute module can use AKS version 1.0 until all container service cmdlets are moved from AzureRm.Compute module.

**Comment by mboersma** [comment]
@hyonholee I understand the concerns. Apologies that I haven't gotten back to this until now.

I'll reformulate this PR so it only adds to a new containerservice folder first, then once we've cleared that hurdle we can move on to step 2 as you describe.

**Comment by mboersma** [comment]
@ravbhatnagar @devigned @dsgouda @lmazuel I reworked this to be only additive, so that we're not concerned it will break existing Compute cmdlets as @hyonholee pointed out.

It is still deficient in that we lack an Operations API endpoint and PATCH endpoints for the main resources. If we could somehow merge it as-is, the AKS team has a plan to rectify that by January.

**Comment by lmazuel** [comment]
@mboersma It's ok for me on the Python side. I pushed the updated Readme for Python, so when this PR is merged the update system is all set. Thanks!

File: specification/containerservices/resource-manager/Microsoft.ContainerService/2017-07-01/containerService.json
Comment (vladbarosan) [inline specification/containerservices/resource-manager/Microsoft.ContainerService/2017-07-01/containerService.json]: @mboersma the semantic validation is failing because it cannot parse the JSON.
https://travis-ci.org/Azure/azure-rest-api-specs/jobs/313171368 

I think the issue is an extra `,` here. Can you have a look please ?

Comment (mboersma) [reply specification/containerservices/resource-manager/Microsoft.ContainerService/2017-07-01/containerService.json]: Thanks for noticing that. Fixed now.

**Comment by vladbarosan** [comment]
@mboersma thanks that fixed it. Before merging:

1. The linter shows 10 errors but you called them out so we dont need to block on them. Did you create an issue for tracking this ? If not please can you create one.

2. There are some model validation issues https://travis-ci.org/Azure/azure-rest-api-specs/jobs/313713796 .

File: specification/containerservices/resource-manager/Microsoft.ContainerService/2017-08-31/examples/ManagedClustersAccessProfilesGet.json
Comment (vladbarosan) [inline specification/containerservices/resource-manager/Microsoft.ContainerService/2017-08-31/examples/ManagedClustersAccessProfilesGet.json]: There is a model validation requires a `roleName` parameter, but you have this. Can you please update with the correct one

File: specification/containerservices/resource-manager/Microsoft.ContainerService/2017-08-31/examples/ManagedClustersAccessProfilesGet.json
Comment (vladbarosan) [inline specification/containerservices/resource-manager/Microsoft.ContainerService/2017-08-31/examples/ManagedClustersAccessProfilesGet.json]: `AccessProfile` is the name of the model, I think this should just be `properties`

**Comment by mboersma** [comment]
@vladbarosan I'll fix up the model validation issues you noted and I'll add the missing x-ms-examples so we end up with 4 errors instead of 10. Then I'll create an issue tracking the remaining errors. I should be able to finish that by Monday. Thanks for your help!

**Comment by vladbarosan** [comment]
@mboersma sounds great!

**Comment by mboersma** [comment]
@vladbarosan I added examples (except for one--`oav` doesn't like the swagger in the locations.json and I'm trying to figure out why). And I have a followup issue at #2129.

**Comment by vladbarosan** [comment]
@mboersma from the linked `Analysis Report` bythe PR bot it seems that in `locations.json` you have an operation `ContainerServices_ListOrchestrators` that does not have an example

**Comment by mboersma** [comment]
@vladbarosan correct, that's the "except for one" I mentioned. When I add the example, it makes `autorest` happy, but when I try to validate it with `oav validate-example` it blows up with a stack trace that suggests it couldn't resolve everything referenced in the locations.json swagger. I'm still trying to find a workaround for this.

**Comment by vladbarosan** [comment]
@mboersma can you share the example you were trying to add via Gist ? I can try it out and see also what might be wrong.

**Comment by mboersma** [comment]
@vladbarosan it was user error (had `x-ms-examples` at the wrong indent level and couldn't see past the irrelevant message that `autorest` was printing.) Should be fixed now.

**Pull request merged**
